lib/spack/docs/build_systems/sconspackage.rst

```diff
@@ -0,0 +1,296 @@
+.. _sconspackage:
+
+------------
+SConsPackage
+------------
+
+Unlike Autotools and CMake, SCons is a general-purpose build system
+that does not rely on Makefiles to build software. SCons is written
+in Python, and handles all building and linking itself.
```

Comment:
Think of Scons as a Python-based "make".  The two should be grouped together.

Reply:
`scons` and `python` should be grouped together or `scons` and `make`? I grouped things based on https://en.wikipedia.org/wiki/List_of_build_automation_software

Reply:
I think that Wikipedia list leaves a lot to be desired, at least for our purposes.  I would classify them as follows:

1. Dependency management systems (of which make was the first): These know that you have to complete tasks A and B before you can run task C.  They keep track of which tasks have been completed as they work their way through the set of tasks to complete.  scons, ant and make are all fundamentally the same idea here.  The difference is how tasks are packaged and run.  In make, tasks are bash commands.  In ant, tasks are Java classes.  In scons, tasks are Python functions.

The fact that scons is written in Python is immaterial here.  Scons is not a "Python build system" in the way that distutils and setuputils are, meaning they are commonly used to build Python packages.  Scons is really a make replacement, for people who think that Python is a better language for Makefiles than GNU make.  The only time I've seen Scons used is in a C++ shop, to build C++.  It turned out to be really slow.  Eventually, it was ripped out and replaced with Makefiles.  It turns out that the GNU make language is (a) fast, and (b) effective due to its functional language lazy-evaluation nature.  Replacing it with a slow, eager-evaluation language was not a good idea.  In the end, everyone's moving to CMake, which makes the lower-level dependency management system (Make / Ninja / Scons / Ant) a little irrelevant, as long as CMake can generate for it.

2. Script-generation systems: CMake, Autotools.  You have to run a configure procedure BEFORE you can run the depenency DAG from (1) above.

3. Meta-builders: Spack


================
@tgamblin We need to get Spack added to that Wikipedia website under"Meta-build".
https://en.wikipedia.org/wiki/List_of_build_automation_software